File: projects/cppsis/src/ScannedColumn.cpp
/*
 * ScannedColumn.cpp
 *
 *  Created on: Mar 9, 2015
 *      Author: peter
 */

#include <iostream>
#include <vector>
#include "ScannedColumn.h"

using namespace std;


ScannedColumn::ScannedColumn() {
	columnIndex = 0;
	size = 0;
	results = new int[size];
	centerPoint = new int[size];
}

ScannedColumn::~ScannedColumn()
{
	columnIndex = 0;
	size = 0;
}

void ScannedColumn::addHeight( int height, int center )
{
	int newsize = size + 1;
	int *newresults = new int[newsize];
	int *newcps = new int[newsize];
	for (int i=0; i < newsize; i++) {
		newresults[i] = results[i];
		newcps[i] = centerPoint[i];
	}
	newresults[size] = height;
	newcps[size] = center;
	delete results;
	delete centerPoint;
	this->size = newsize;
	this->results = newresults;
	this->centerPoint = newcps;
	// printString();
}

void ScannedColumn::setColumnIndex( int index )
{
	columnIndex = index;
}

int ScannedColumn::getColumnIndex()
{
	return columnIndex;
}

int ScannedColumn::getHeight(int index)
{
	if (index < this->size) {
		return results[index];
	} else {
		return 0;
	}
}

int ScannedColumn::getCenter(int index)
{
	if (index < this->size) {
		return centerPoint[index];
	} else {
		return 0;
	}
}

int ScannedColumn::getCount()
{
	return this->size;
}

void ScannedColumn::printString(void) {
	cout << this->columnIndex << ":";
	cout <<"[";
	for (int i=0; i < this->size; i++) {
		if (i > 0) { cout << ","; }
		cout << this->results[i];
	}
	cout << "] - ";
	cout <<"[";
	for (int i=0; i < this->size; i++) {
		if (i > 0) { cout << ","; }
		cout << this->centerPoint[i];
	}
	cout << "]";
	cout << endl;
}

File: projects/JavaScanner/src/org/woolfel/cv/ScannedColumn.java
package org.woolfel.cv;

import java.lang.Integer;
import java.util.*;

/**
 * Encapsulates the results of scanning an image for black lines
 * 
 * @author peter lin
 *
 */
public class ScannedColumn {

	private int columnIndex = 0;
	private int[] results = new int[0];
	private int[] centerPoint = new int[0];
	
	public ScannedColumn() {
	}

	public void addHeight(int height, int center) {
		int[] newrs = new int[results.length + 1];
		int[] newcp = new int[centerPoint.length + 1];
		System.arraycopy(results, 0, newrs, 0, results.length);
		System.arraycopy(centerPoint, 0, newcp, 0, centerPoint.length);
		newrs[results.length] = height;
		newcp[results.length] = center;
		results = newrs;
		centerPoint = newcp;
	}
	
	public int count() {
		return results.length;
	}
	
	public void setColumnIndex(int in) {
		this.columnIndex = in;
	}
	
	public int getColumnIndex() {
		return this.columnIndex;
	}
	
	public int getHeight(int index) {
		if (index >= this.results.length) {
			return 0;
		} else {
			return this.results[index];
		}
	}
	
	public int getCenter(int index) {
		if (index >= this.centerPoint.length) {
			return 0;
		} else {
			return this.centerPoint[index];
		}
	}
	
	/**
	 * Returns an array: height, center
	 * @param index
	 * @return
	 */
	public int[] getValues(int index) {
		int[] data = new int[2];
		data[0] = this.results[index];
		data[1] = this.centerPoint[index];
		return data;
	}
	
	public boolean isEmpty() {
		return this.results.length == 0;
	}
	
	public String toString() {
		return columnIndex + ":"+ arrayString(this.results) + " - " + arrayString(centerPoint);
	}
	
	private String arrayString(int[] array) {
		StringBuffer buf = new StringBuffer();
		buf.append("[");
		for (int i=0; i < array.length; i++) {
			if (i > 0) {
				buf.append(",");
			}
			buf.append(array[i]);
		}
		buf.append("]");
		return buf.toString();
	}
}
